File: bindings/dotnet/Kriya/Validation.cs
using System.Text.Json;
using System.Text.Json.Nodes;

namespace Kriya;

/// <summary>One problem found validating params against their schemas.</summary>
public sealed record ValidationIssue(string Path, string Message);

/// <summary>
/// Runtime validation of action parameters against their declared schemas — what stops an agent (or
/// a buggy caller) invoking a handler with the wrong shape. Types, enums, required-ness, array
/// element and object property types. Unknown params are ignored (forward-compatible). Faithful port
/// of kriya-core's <c>validate.ts</c>.
/// </summary>
public static class Validation
{
    private static string TypeOf(JsonNode? node)
    {
        if (node is null) return "null";
        return node.GetValueKind() switch
        {
            JsonValueKind.String => "string",
            JsonValueKind.Number => "number",
            JsonValueKind.True or JsonValueKind.False => "boolean",
            JsonValueKind.Array => "array",
            JsonValueKind.Object => "object",
            JsonValueKind.Null => "null",
            _ => "object",
        };
    }

    private static void Check(JsonNode? value, ParameterSchema schema, string path, List<ValidationIssue> issues)
    {
        var actual = TypeOf(value);
        var expected = JsonSchema.TypeName(schema.Type);
        if (actual != expected)
        {
            issues.Add(new ValidationIssue(path, $"expected {expected}, got {actual}"));
            return; // type wrong — deeper checks would be noise
        }

        if (schema.Enum is not null)
        {
            var asJson = value?.ToJsonString();
            var ok = schema.Enum.Any(e => JsonValue.Create(e)?.ToJsonString() == asJson);
            if (!ok)
                issues.Add(new ValidationIssue(path, $"value {asJson} is not one of the allowed values"));
        }

        if (schema.Type == ParamType.Array && schema.Items is not null && value is JsonArray arr)
        {
            for (var i = 0; i < arr.Count; i++)
                Check(arr[i], schema.Items, $"{path}[{i}]", issues);
        }

        if (schema.Type == ParamType.Object && schema.Properties is not null && value is JsonObject obj)
        {
            foreach (var (key, prop) in schema.Properties)
            {
                if (!obj.ContainsKey(key))
                {
                    if (prop.Required) issues.Add(new ValidationIssue($"{path}.{key}", "required"));
                    continue;
                }
                Check(obj[key], prop, $"{path}.{key}", issues);
            }
        }
    }

    /// <summary>Validate a params object against a map of named parameter schemas.</summary>
    public static List<ValidationIssue> ValidateParams(JsonObject parameters, IReadOnlyDictionary<string, ParameterSchema> schemas)
    {
        var issues = new List<ValidationIssue>();
        foreach (var (name, schema) in schemas)
        {
            if (!parameters.ContainsKey(name))
            {
                if (schema.Required) issues.Add(new ValidationIssue(name, "required"));
                continue;
            }
            Check(parameters[name], schema, name, issues);
        }
        return issues;
    }

    /// <summary>Format issues into a single human/agent-readable string.</summary>
    public static string FormatIssues(IEnumerable<ValidationIssue> issues) =>
        string.Join("; ", issues.Select(i => $"{i.Path}: {i.Message}"));
}

File: bindings/java/src/main/java/com/kriyanative/kriya/Validation.java
package com.kriyanative.kriya;

import java.util.ArrayList;
import java.util.List;
import java.util.Map;

/**
 * Runtime validation of action parameters against their declared schemas — what stops an agent (or a
 * buggy caller) invoking a handler with the wrong shape. Types, enums, required-ness, array element
 * and object property types. Unknown params are ignored (forward-compatible). Faithful port of
 * kriya-core's {@code validate.ts}.
 */
public final class Validation {
    private Validation() {}

    private static String typeOf(Object node) {
        if (node == null) return "null";
        if (node instanceof String) return "string";
        if (node instanceof Boolean) return "boolean";
        if (node instanceof Number) return "number";
        if (node instanceof List) return "array";
        if (node instanceof Map) return "object";
        return "object";
    }

    @SuppressWarnings("unchecked")
    private static void check(Object value, ParameterSchema schema, String path, List<ValidationIssue> issues) {
        String actual = typeOf(value);
        String expected = JsonSchema.typeName(schema.type);
        if (!actual.equals(expected)) {
            issues.add(new ValidationIssue(path, "expected " + expected + ", got " + actual));
            return; // type wrong — deeper checks would be noise
        }

        if (schema.enumValues != null) {
            String asJson = Json.stringify(value);
            boolean ok = false;
            for (Object e : schema.enumValues) {
                if (Json.stringify(e).equals(asJson)) { ok = true; break; }
            }
            if (!ok) issues.add(new ValidationIssue(path, "value " + asJson + " is not one of the allowed values"));
        }

        if (schema.type == ParamType.ARRAY && schema.items != null && value instanceof List) {
            List<Object> arr = (List<Object>) value;
            for (int i = 0; i < arr.size(); i++) {
                check(arr.get(i), schema.items, path + "[" + i + "]", issues);
            }
        }

        if (schema.type == ParamType.OBJECT && schema.properties != null && value instanceof Map) {
            Map<String, Object> obj = (Map<String, Object>) value;
            for (Map.Entry<String, ParameterSchema> e : schema.properties.entrySet()) {
                String key = e.getKey();
                ParameterSchema prop = e.getValue();
                if (!obj.containsKey(key)) {
                    if (prop.required) issues.add(new ValidationIssue(path + "." + key, "required"));
                    continue;
                }
                check(obj.get(key), prop, path + "." + key, issues);
            }
        }
    }

    /** Validate a params object against a map of named parameter schemas. */
    public static List<ValidationIssue> validateParams(Map<String, Object> parameters, Map<String, ParameterSchema> schemas) {
        List<ValidationIssue> issues = new ArrayList<>();
        for (Map.Entry<String, ParameterSchema> e : schemas.entrySet()) {
            String name = e.getKey();
            ParameterSchema schema = e.getValue();
            if (parameters == null || !parameters.containsKey(name)) {
                if (schema.required) issues.add(new ValidationIssue(name, "required"));
                continue;
            }
            check(parameters.get(name), schema, name, issues);
        }
        return issues;
    }

    /** Format issues into a single human/agent-readable string. */
    public static String formatIssues(List<ValidationIssue> issues) {
        StringBuilder sb = new StringBuilder();
        for (int i = 0; i < issues.size(); i++) {
            if (i > 0) sb.append("; ");
            sb.append(issues.get(i).path).append(": ").append(issues.get(i).message);
        }
        return sb.toString();
    }
}
